tests/factories.py

```diff
@@ -243,6 +248,145 @@ def create_release_request(workspace, user=None, **kwargs):
+    )
+    if status == RequestStatus.WITHDRAWN:
+        assert (
+            withdrawn_after is not None
+        ), "pass withdrawn_after to decide when to withdraw"
+
+    # Get a default checker if one was not provided
+    # This is the checker who does the state transitions (approved/released/returned/rejected)
+    # It is not necessarily the same checker who reviews files.
+    # `request_file()` can be called with an optional list of checkers; if it is
+    # None, the same default checkers will be used for reviewing files (and the first one will
+    # be the one that does the other state transitions.)
+    if checker:
+        file_reviewers = [checker, get_default_output_checkers()[1]]
+    else:
+        file_reviewers = get_default_output_checkers()
+        checker = file_reviewers[0]
+
+    request = create_release_request(
+        workspace, author, status=RequestStatus.PENDING, **kwargs
+    )
+
+    # add all files
+    for add_file in files:
+        add_file(request)
+
+    request = refresh_release_request(request)
+    if status == RequestStatus.PENDING:
+        return request
+
+    if status == RequestStatus.WITHDRAWN and withdrawn_after == RequestStatus.PENDING:
+        bll.set_status(request, RequestStatus.WITHDRAWN, author)
+        return refresh_release_request(request)
+
+    bll.submit_request(request, author)
+    request = refresh_release_request(request)
+
+    if status == RequestStatus.SUBMITTED:
+        return request
+
+    # If there are output files, get the usernames of all file reviewers
+    # so we can complete reviews with the correct checkers.
+    # Note that it is possible to review a request with no output files
+    # (potentially before returning it to the reviewer so they can add some).
+    # Approving or releasing requests with no output files is not allowed.
+    if request.output_files():
+        file_reviewers = [
+            User(username, output_checker=True)
+            for username in list(request.output_files().values())[0].reviews.keys()
+        ]
+
+    if status == RequestStatus.PARTIALLY_REVIEWED:
+        complete_independent_review(request, file_reviewers[0])
+        return refresh_release_request(request)
+    # all other state require completed reviews.
+    complete_independent_review(request, *file_reviewers)
+    request = refresh_release_request(request)
+
+    if status == RequestStatus.REVIEWED:
+        return request
+
+    if status in [RequestStatus.RETURNED, RequestStatus.WITHDRAWN]:
+        bll.set_status(request, RequestStatus.RETURNED, checker)
+        request = refresh_release_request(request)
+
+        if (
+            status == RequestStatus.WITHDRAWN
+            and withdrawn_after == RequestStatus.RETURNED
+        ):
+            bll.set_status(request, RequestStatus.WITHDRAWN, author)
+            return refresh_release_request(request)
+        else:
+            return request
```

Comment:
I like this guard-clause type pattern of returning, but we can get rid of some of these else clauses to tidy up 
```suggestion
    if status in [RequestStatus.RETURNED, RequestStatus.WITHDRAWN]:
        bll.set_status(request, RequestStatus.RETURNED, checker)
        request = refresh_release_request(request)

        if (
            status == RequestStatus.WITHDRAWN
            and withdrawn_after == RequestStatus.RETURNED
        ):
            bll.set_status(request, RequestStatus.WITHDRAWN, author)
            return refresh_release_request(request)
  
        return request
```

Reply:
actually I'd like it more if the branches were flipped, but we're really into code golf now!
```suggestion
    if status in [RequestStatus.RETURNED, RequestStatus.WITHDRAWN]:
        bll.set_status(request, RequestStatus.RETURNED, checker)
        request = refresh_release_request(request)

        if not (
            status == RequestStatus.WITHDRAWN
            and withdrawn_after == RequestStatus.RETURNED
        ):
            return request

        bll.set_status(request, RequestStatus.WITHDRAWN, author)
        return refresh_release_request(request)
```